presto-orc/src/main/java/com/facebook/presto/orc/metadata/ExceptionWrappingMetadataReader.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.orc.metadata;
+
+import com.facebook.presto.orc.OrcCorruptionException;
+import com.facebook.presto.orc.OrcDataSourceId;
+import com.facebook.presto.orc.metadata.PostScript.HiveWriterVersion;
+import com.facebook.presto.orc.metadata.statistics.HiveBloomFilter;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+public class ExceptionWrappingMetadataReader
+        implements MetadataReader
+{
+    private final OrcDataSourceId orcDataSourceId;
+    private final MetadataReader delegate;
+
+    public static ExceptionWrappingMetadataReader createExceptionWrappingMetadataReader(OrcDataSourceId orcDataSourceId, MetadataReader delegate)
+    {
+        if (delegate instanceof ExceptionWrappingMetadataReader) {
```

Comment:
This should probably be `checkArgument` and fail if it's already wrapped. The factory method can be deleted after this.